### PR TITLE
Use detours instead of replacing BuildingManager

### PR DESCRIPTION
**Pull request by bloodypenguin**

### CODE REVIEW

File: BuildingThemes/DetoursHolder.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        private static ushort blockId; //TODO(earalov): maybe ZoneBlock
+        
+
+        //this is detoured version of BuildingManger#GetRandomBuildingInfo method. Note, that it's an instance method. It's better because this way all registers will be expected to have the same values
+        //as in original methods
+        public BuildingInfo GetRandomBuildingInfo(ref Randomizer r, ItemClass.Service service, ItemClass.SubService subService, ItemClass.Level level, int width, int length, BuildingInfo.ZoningMode zoningMode)
+        {
+            Debug.Log("Building Themes: Detoured GetRandomBuildingInfo was called.");
+            //this part is the same as in original method
+            var buildingManager = (BuildingManager)Convert.ChangeType(this, typeof(BuildingManager)); //I love this hack :)
```
Comment (bloodypenguin) [inline BuildingThemes/DetoursHolder.cs]: This is how to get ``this`. Compare it to original method

File: BuildingThemes/DetoursHolder.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    class DetoursHolder
+    {
+
+        //we'll use this variable to pass block to GetRandomBuildingInfo method
+        private static ushort blockId; //TODO(earalov): maybe ZoneBlock
+        
+
+        //this is detoured version of BuildingManger#GetRandomBuildingInfo method. Note, that it's an instance method. It's better because this way all registers will be expected to have the same values
+        //as in original methods
+        public BuildingInfo GetRandomBuildingInfo(ref Randomizer r, ItemClass.Service service, ItemClass.SubService subService, ItemClass.Level level, int width, int length, BuildingInfo.ZoningMode zoningMode)
```
Comment (bloodypenguin) [inline BuildingThemes/DetoursHolder.cs:21]: Note that it's instance method.

File: BuildingThemes/BuildingThemesMod.cs [hunk 3]
```diff
@@ -51,47 +53,13 @@ private string GetCurrentEnvironment()
 
         private void ReplaceBuildingManager()
```
Comment (bloodypenguin) [inline BuildingThemes/BuildingThemesMod.cs:54]: It's better be called from `OnCreated` method

**Comment by boformer** [comment]
This is exactly what I had in mind, but I didn't have the C# knowledge. Nice work!

**Comment by boformer** [comment]
ready for merge? You can do it yourself when it is ready, you are now a collaborator.

**Pull request merged**